src/dataflow-types/src/client/controller/storage.rs

```diff
@@ -485,38 +486,71 @@ where
-                .state
-                .persist_handles
-                .get_mut(&id)
-                .expect("unknown collection id");
+            let (total_updates, new_upper) = updates_by_id
+                .entry(id)
+                .or_insert_with(|| (Vec::new(), T::minimum()));
+            total_updates.push(updates);
+            new_upper.join_assign(&batch_upper);
+        }
+
+        let mut appends_by_id = HashMap::new();
+        for (id, (updates, upper)) in updates_by_id {
+            let current_upper = self.collection(id)?.write_frontier.frontier().to_owned();
+            appends_by_id.insert(id, (updates.into_iter().flatten(), current_upper, upper));
+        }
+
+        let futs = FuturesUnordered::new();
+
+        // We cannot iterate through the updates and then set off a persist call
+        // on the write handle because we cannot mutably borrow the write handle
+        // multiple times.
+        //
+        // Instead, we first group the update by ID above and then iterate
+        // through all available write handles and see if there are any updates
+        // for it. If yes, we send them all in one go.
+        for (id, persist_handle) in self.state.persist_handles.iter_mut() {
+            let (updates, upper, new_upper) = match appends_by_id.remove(id) {
+                Some(updates) => updates,
+                None => continue,
+            };
+
+            let new_upper = Antichain::from_elem(new_upper);
 
             let updates = updates
                 .into_iter()
                 .map(|u| ((SourceData(Ok(u.row)), ()), u.timestamp, u.diff));
 
-            handles
-                .write
-                .compare_and_append(updates, upper.clone(), new_upper.clone())
-                .await
-                .expect("cannot append updates")
-                .expect("cannot append updates")
-                .or(Err(StorageError::InvalidUpper(id)))?;
-
-            let mut change_batch = ChangeBatch::new();
-            change_batch.extend(new_upper.iter().cloned().map(|t| (t, 1)));
-            change_batch.extend(upper.iter().cloned().map(|t| (t, -1)));
-            self.update_write_frontiers(&[(id, change_batch)]).await?;
+            let write = &mut persist_handle.write;
+
+            futs.push(async move {
+                write
+                    .compare_and_append(updates, upper.clone(), new_upper.clone())
+                    .await
+                    .expect("cannot append updates")
+                    .expect("cannot append updates")
+                    .or(Err(StorageError::InvalidUpper(*id)))?;
+
+                let mut change_batch = ChangeBatch::new();
+                change_batch.extend(new_upper.iter().cloned().map(|t| (t, 1)));
+                change_batch.extend(upper.iter().cloned().map(|t| (t, -1)));
+
+                Ok::<_, StorageError>((*id, change_batch))
+            })
         }
+
+        let change_batches = futs.try_collect::<Vec<_>>().await?;
```

Comment:
can anything bad happen if one of these futures is cancelled in the middle of an await? for example, one could be in the middle of `compare_and_append`, but another finishes with an error, and the first is dropped

Reply:
I _think_ it should be fine, at least I don't think it makes things worse. Before, a `compare_and_append` could be cancelled right in the middle of sth if the future that is returned from `append()` is cancelled.